### ISSUE TITLE
attribute value constraint and processing disagreement

### ISSUE THREAD

**Issue by sydb** [TEIC/Stylesheets#353]
In creating schemas via ODD we use lots of attributes that are declared as teidata.name (e.g. `@key` and `@ident`). RELAX NG permits leading and trailing whitespace. However odds/odd2odd.xsl does not strip leading and trailing whitespace.

Thus
~~~~~xml
<memberOf key="att.typed"/>
~~~~~
is valid and works fine, but
~~~~~xml
<memberOf key=" att.typed "/>
~~~~~
is valid but silently fails: in the output schema the element being specified does not actually get the `@type` and `@subtype` attributes.

Seems to me we should either
1. Shore up the schema so that it objects if there is whitespace in the value of one of these kinds of attriubtes, perhaps via an added Schematron rule.
1. Test for spaces in the processing and issue a warning or error if they are present.
1. Test to see if the `<memberOf>` (or `<dataRef>` or whatever) found something via the provided `@key`, and issue a warning if not found. (I haven’t thought this through at all, but I suspect would be the best way to go for `@key` but would not work at all for `@ident`.) 
1. Consistently use normalize-space() on these values in the processing.

I don’t have a preference at the moment. Also seems to me to be pretty low priority, as it has probably been this way for well over a decade and no one has complained.

**Comment by lb42** [comment]
Option 4 plz.


Next.

reluctantly using Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by martindholmes** [comment]
All four options please. :-)

**Comment by jamescummings** [comment]
Definitely 4

**Comment by jamescummings** [comment]
Are you allowed to have multiple values? If so we should process with that in mind. If not then why not strip them with translate or something rather than normalise space? Not saying that is less scary.

**Comment by sydb** [comment]
No, @jamescummings, the values I am talking about are singular. (There are others that are multiple, like `moduleRef/@include`.) And sure, there are other ways to do this.

**Comment by stuartyeates** [comment]
4

Teidata.name is defined as whitespace-free in the text, so this is just
checking an existing constraint.

Cheers
Stuart

**Comment by sydb** [comment]
Set to “pending” to verify that the original problem really is a problem, i.e. that `" att.typed "` is valid and does produce processing errors.

**Comment by sydb** [comment]
Assigning myself to perform the verification described [above](#issuecomment-637597160).

**Comment by sydb** [comment]
Verified. [This little ODD](https://github.com/TEIC/Stylesheets/files/9537822/tst_353.odd.txt) demonstrates the problem. It is valid against tei_all and p5odds (and _almost_ valid against tei_customization). But the added elements show up as
~~~
tst_353_one =
  element one {
    xsd:token { pattern = "[^\p{C}\p{Z}]+" },
    tst_353_att.global.attributes,
    tst_353_att.typed.attributes,
    empty
  }

tst_353_two =
  element two {
    xsd:token { pattern = "[^\p{C}\p{Z}]+" },
    tst_353_att.global.attributes,
    empty
  }
~~~
in the RNC. (Note the missing “tst_353_att.typed.attributes” for the definition of `<tst:two>`.)

**Comment by sydb** [comment]
Since no one else has picked this up in the last 20 months, I have re-assigned it to myself.
Sigh.
